resctrl/utils.go

```diff
@@ -96,47 +106,53 @@ func prepareMonitoringGroup(containerName string, getContainerPids func() ([]str
 	}
 
 	if len(pids) == 0 {
-		return "", fmt.Errorf("couldn't obtain %q container pids, there is no pids in cgroup", containerName)
+		return "", fmt.Errorf("couldn't obtain %q container pids: there is no pids in cgroup", containerName)
 	}
 
-	path, err := findControlGroup(pids)
+	controlGroupPath, err := findGroup(rootResctrl, pids, true, false)
 	if err != nil {
-		return "", fmt.Errorf("couldn't find control group matching %q container: %v", containerName, err)
+		return "", fmt.Errorf("%q %q: %q", noControlGroupFoundError, containerName, err)
 	}
-
-	if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
-		containerName = containerName[1:]
+	if controlGroupPath == "" {
+		return "", fmt.Errorf("%q %q", noControlGroupFoundError, containerName)
 	}
 
-	properContainerName := strings.Replace(containerName, "/", "-", -1)
-	monGroupPath := filepath.Join(path, monitoringGroupDir, properContainerName)
-
-	// Create new mon_group if not exists.
-	err = os.MkdirAll(monGroupPath, os.ModePerm)
+	// Check if there is any monitoring group.
+	monGroupPath, err := findGroup(filepath.Join(controlGroupPath, monGroupsDirName), pids, false, true)
 	if err != nil {
-		return "", fmt.Errorf("couldn't create monitoring group directory for %q container: %w", containerName, err)
+		return "", fmt.Errorf("couldn't find monitoring group matching %q container: %v", containerName, err)
 	}
 
-	for _, pid := range pids {
-		threadFiles, err := ioutil.ReadDir(filepath.Join(processPath, pid, processTask))
-		if err != nil {
-			return "", err
+	// Prepare new one if not exists.
+	if monGroupPath == "" {
+		if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
+			containerName = containerName[1:]
 		}
 
-		processThreads, err := getAllProcessThreads(threadFiles)
+		properContainerName := fmt.Sprintf("%s-%s", monGroupPrefix, strings.Replace(containerName, "/", "-", -1))
+		monGroupPath = filepath.Join(controlGroupPath, monitoringGroupDir, properContainerName)
+
+		err = os.MkdirAll(monGroupPath, os.ModePerm)
 		if err != nil {
-			return "", err
+			return "", fmt.Errorf("couldn't create monitoring group directory for %q container: %w", containerName, err)
 		}
-		for _, thread := range processThreads {
-			err = intelrdt.WriteIntelRdtTasks(monGroupPath, thread)
+
+		for _, pid := range pids {
+			processThreads, err := getAllProcessThreads(filepath.Join(processPath, pid, processTask))
 			if err != nil {
-				secondError := os.Remove(monGroupPath)
-				if secondError != nil {
-					return "", fmt.Errorf(
-						"coudn't assign pids to %q container monitoring group: %w \n couldn't clear %q monitoring group: %v",
-						containerName, err, containerName, secondError)
+				return "", err
+			}
+			for _, thread := range processThreads {
+				err = intelrdt.WriteIntelRdtTasks(monGroupPath, thread)
+				if err != nil {
```

Comment:
what if just of the pids disapperead but rest are ok? why are you removing all the group

Reply:
8c734a03c1bc0e803b5b5cbbca03882ea0135bb5 will be sufficient for you?

Reply:
That will be a race condition which `--resctrl_interval` fix.

Reply:
I think that we shouldn't monitor something that cannot be set by us.